modular/downloader/download_task.go

```diff
@@ -207,34 +244,56 @@ func (d *DownloadModular) PreDownloadPiece(ctx context.Context, downloadPieceTas
 		return nil
 	}
 	if downloadPieceTask.GetEnableCheck() {
-		// if it is a request from client, the task handler is the primary SP of the sp
-		if d.baseApp.OperatorAddress() == bucketPrimarySp.OperatorAddress {
-			log.CtxDebugw(ctx, "downloading from primary SP, checking quota")
-			checkQuotaTime := time.Now()
-			if err := d.baseApp.GfSpDB().CheckQuotaAndAddReadRecord(
-				&spdb.ReadRecord{
-					BucketID:        downloadPieceTask.GetBucketInfo().Id.Uint64(),
-					ObjectID:        downloadPieceTask.GetObjectInfo().Id.Uint64(),
-					UserAddress:     downloadPieceTask.GetUserAddress(),
-					BucketName:      downloadPieceTask.GetBucketInfo().GetBucketName(),
-					ObjectName:      downloadPieceTask.GetObjectInfo().GetObjectName(),
-					ReadSize:        downloadPieceTask.GetTotalSize(),
-					ReadTimestampUs: sqldb.GetCurrentTimestampUs(),
-				},
-				&spdb.BucketQuota{
-					ReadQuotaSize: downloadPieceTask.GetBucketInfo().GetChargedReadQuota() + d.bucketFreeQuota,
-				},
-			); err != nil {
-				metrics.PerfGetObjectTimeHistogram.WithLabelValues("get_object_check_quota_time").Observe(time.Since(checkQuotaTime).Seconds())
-				log.CtxErrorw(ctx, "failed to check bucket quota", "error", err)
-				if errors.Is(err, sqldb.ErrCheckQuotaEnough) {
-					return ErrExceedBucketQuota
-				}
-				// ignore the access db error, it is the system's inner error, will be let the request go.
+		checkQuotaTime := time.Now()
+
+		bucketID := downloadPieceTask.GetBucketInfo().Id.Uint64()
+		bucketName := downloadPieceTask.GetBucketInfo().GetBucketName()
+		bucketTraffic, err = d.baseApp.GfSpDB().GetBucketTraffic(downloadPieceTask.GetBucketInfo().Id.Uint64())
+		if err != nil && !errors.Is(err, gorm.ErrRecordNotFound) {
+			return err
+		}
+		// init the bucket traffic table when checking quota for the first time
+		if bucketTraffic == nil {
+			freeQuotaSize, err = d.baseApp.Consensus().QuerySPFreeQuota(ctx, d.baseApp.OperatorAddress())
+			if err != nil {
+				return ErrConsensus
 			}
+			log.CtxDebugw(ctx, "finish init bucket traffic table", "charged quota:", downloadPieceTask.GetBucketInfo().GetChargedReadQuota(),
```

Comment:
Refine "charged quota:" to "charged_quota"

Reply:
fixed